ledger/complete/mtrie/trie/trie.go

```diff
@@ -305,46 +341,62 @@ func (mt *MTrie) proofs(head *node.Node, paths []ledger.Path, proofs []*ledger.T
 	// partition step to quick sort the paths:
 	// lpaths contains all paths that have `0` at the partitionIndex
 	// rpaths contains all paths that have `1` at the partitionIndex
-	heightIndex := mt.Height() - head.Height() // distance to the tree root
-	partitionIndex := utils.SplitTrieProofsByPath(paths, proofs, heightIndex)
+	depth := mt.Height() - head.Height() // distance to the tree root
+	partitionIndex := utils.SplitTrieProofsByPath(paths, proofs, depth)
 	lpaths, rpaths := paths[:partitionIndex], paths[partitionIndex:]
 	lproofs, rproofs := proofs[:partitionIndex], proofs[partitionIndex:]
 
-	wg := sync.WaitGroup{}
-	parallelRecursionThreshold := 128 // thresold to avoid the parallelization going too deep in the recursion
+	parallelRecursionThreshold := 128 // threshold to avoid the parallelization going too deep in the recursion
+	if len(lpaths) <= parallelRecursionThreshold {
+		// runtime optimization: below the parallelRecursionThreshold, we proceed single-threaded
+		addSubtrieHashToProofs(head.RightChild(), depth, lproofs)
+		mt.proofs(head.LeftChild(), lpaths, lproofs)
 
-	if rChild := head.RightChild(); rChild != nil { // TODO: is that a sanity check?
-		nodeHash := rChild.Hash()
-		isDef := bytes.Equal(nodeHash, common.GetDefaultHashForHeight(rChild.Height())) // TODO: why not rChild.RegisterCount != 0?
-		if !isDef {                                                                     // in proofs, we only provide non-default value hashes
-			for _, p := range lproofs {
-				utils.SetBit(p.Flags, heightIndex)
-				p.Interims = append(p.Interims, nodeHash)
-			}
-		}
-	}
-	if len(lpaths) > parallelRecursionThreshold {
+		addSubtrieHashToProofs(head.LeftChild(), depth, rproofs)
+		mt.proofs(head.RightChild(), rpaths, rproofs)
+	} else {
+		wg := sync.WaitGroup{}
 		wg.Add(1)
 		go func() {
+			addSubtrieHashToProofs(head.RightChild(), depth, lproofs)
 			mt.proofs(head.LeftChild(), lpaths, lproofs)
 			wg.Done()
 		}()
-	} else {
-		mt.proofs(head.LeftChild(), lpaths, lproofs)
+
+		addSubtrieHashToProofs(head.LeftChild(), depth, rproofs)
+		mt.proofs(head.RightChild(), rpaths, rproofs)
+		wg.Wait()
 	}
+}
 
-	if lChild := head.LeftChild(); lChild != nil {
-		nodeHash := lChild.Hash()
-		isDef := bytes.Equal(nodeHash, common.GetDefaultHashForHeight(lChild.Height()))
-		if !isDef { // in proofs, we only provide non-default value hashes
-			for _, p := range rproofs {
-				utils.SetBit(p.Flags, heightIndex)
-				p.Interims = append(p.Interims, nodeHash)
-			}
+// addSubtrieHashToProofs inspects the other subtrie and adds its root hash
+// to the proofs, if the trie contains non-empty registers (i.e. the
+// otherSubtrie has a non-default hash).
+func addSubtrieHashToProofs(otherSubtrie *node.Node, depth int, proofs []*ledger.TrieProof) {
```

Comment:
👌

Reply:
I suggest to rename `otherSubtrie` to `siblingTrie` as used in the trees vocabulary.

Reply:
done in https://github.com/onflow/flow-go/pull/562/commits/628dbf1df7dead82fdc9f0096c8c95a615708e4b